# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np

In [2]:
#os.getcwd()

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [3]:
path="C:\\Users\\Admin\\Dropbox\\Rory\\01 - Development\\Data Science\\Springboard\\Career Track\\06 Applying the Data Science Method\\Step 02 - Data Wrangling"
os.chdir(path)

In [4]:
#os.listdir()

In [5]:
ski3 = pd.read_csv('step3_output.csv', index_col=0)
#ski3.head()

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [6]:
state = ski3['state']
df_mod1 = pd.concat([ski3.drop(columns = 'state', axis=1), pd.get_dummies(state)], axis=1)
#ski3.head()

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [7]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing as pp

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df_mod1.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df_mod1.AdultWeekend 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = pp.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [8]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [9]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [10]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test

y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [11]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

In [12]:
from sklearn.metrics import explained_variance_score
from sklearn.metrics import mean_absolute_error
evs = explained_variance_score(y_test, y_pred)
mbe = mean_absolute_error(y_test, y_pred)
print(evs, mbe)

-7.942936358576415e+19 40626581916.03664


**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [13]:
lm.intercept_

-10350721448.241615

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [14]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values('Coefficient', ascending=False).head(10)

Coefficient
fastEight     3.727318e+13
total_chairs  1.060870e+13
fastSixes     8.553480e+12
double        5.757596e+12
surface       5.442102e+12
trams         5.339624e+12
triple        4.813981e+12
quad          2.692807e+12
fastQuads     2.269215e+12
New York      1.592396e+11

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [15]:
X = ski3.drop(['Name','AdultWeekend', 'state'], axis=1)
y = ski3.AdultWeekend 
scaler = pp.StandardScaler().fit(X) 
X_scaled=scaler.transform(X) 

In [16]:
y = y.ravel() 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [17]:
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

In [18]:
y_pred = model.predict(X_test)

In [19]:
evs = explained_variance_score(y_test, y_pred)
mbe = mean_absolute_error(y_test, y_pred)
print(evs, mbe)

0.7549328209296664 6.243902655590581


In [20]:
lm.intercept_

57.87249727207076

In [21]:
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values('Coefficient', ascending=False).head(10)

Coefficient
AdultWeekday         10.576186
vertical_drop         2.854249
clusters              2.512542
summit_elev           2.471256
quad                  1.326506
SkiableTerrain_ac     1.323618
averageSnowfall       1.084366
triple                1.034691
TerrainParks          0.993711
total_chairs          0.648883

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

Rory Note: 'base_elev' was already dropped from dataset during EDA stage due to high correlation.

In [22]:
X = ski3.drop(['Name','AdultWeekend', 'state', 'summit_elev'], axis=1)
y = ski3.AdultWeekend 
scaler = pp.StandardScaler().fit(X) 
X_scaled=scaler.transform(X)

In [23]:
y = y.ravel() 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

In [24]:
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

In [25]:
y_pred = model.predict(X_test)

In [26]:
evs = explained_variance_score(y_test, y_pred)
mbe = mean_absolute_error(y_test, y_pred)
print(evs, mbe)

0.7409742835382752 6.551153073502763


In [27]:
lm.intercept_

57.86780970728353

In [28]:
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values('Coefficient', ascending=False).head(10)

Coefficient
AdultWeekday         10.514102
vertical_drop         2.632334
SkiableTerrain_ac     1.889379
clusters              1.872010
averageSnowfall       1.516254
quad                  1.242833
TerrainParks          1.107535
NightSkiing_ac        0.792426
total_chairs          0.704035
triple                0.638038

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | -7.942936358576415e+19 | 40626581916.03664 |-|
| Model 2. | 0.7549328209296664 | 6.243902655590581 |'state'|
| Model 3. | 0.7409742835382752 | 6.551153073502763 |'state','summit_elev','base_elev'|

Model Selection:

Model 2 and Model 3 both have high explained variance which means they are the better fitting models. Due to the correlation that may be present in Model 2 between certain explantory variables, it is recommended to go with Model 3. 

This model shows that there are physical and weather features that allow certain parks to charge higher prices such as Vertical Drop, size of skiing area and level of snowfall. However proceeding this, the number of chairs by type are seen to be features that skiiers are willing to pay extra for with quads, triples and overall number of chairs being in the top ten. This model roughly shows that, depending on chair type, the park could increase Weekend pricing by $0.63 to $1.24 per additional chair added to the resort.